### PR TITLE
Test on PHP 8.1 with GitHub Actions

### PR DESCRIPTION
**Pull request by davidrans**
And drop Travis CI and PHP 7 support

### CODE REVIEW

**Comment by davidrans** [comment]
dev_block ✋

File: .github/workflows/php.yml [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+
+    - name: Install PHP with extensions
+      uses: shivammathur/setup-php@v2
+      with:
+        php-version: ${{ matrix.php-version }}
+        extensions: memcached
+        coverage: none
+
+    - name: Setup Memcached server
+      uses: niden/actions-memcached@v7
```
Comment (mlahargou) [inline .github/workflows/php.yml:40]: Does this replace `services:memcached` in the travis yaml?

File: .github/workflows/php.yml [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+
+    - name: Validate composer.json and composer.lock
+      run: composer validate --strict
+
+    - name: Cache Composer packages
```
Comment (mlahargou) [inline .github/workflows/php.yml:45]: Mind explaining what this section does? Looks like we cache the composer lock file on a per `os` basis but it looks like only run on one `os`.

File: .github/workflows/php.yml [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+    runs-on: ubuntu-latest
+
+    strategy:
+      matrix:
+        php-version:
+          - "8.0"
+          - "8.1"
+
+        include:
+          - os: ubuntu-latest
```
Comment (mlahargou) [inline .github/workflows/php.yml:21]: I know with our docker scripts we try to stay away from using `latest`. Do we want to specify a version here?

File: .github/workflows/php.yml [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+          ${{ runner.os }}-php-
+
+    - name: Install dependencies
+      run: composer install --prefer-dist --no-progress
+
+    - name: Run tests with phpunit
+      run: ./vendor/bin/phpunit tests
+
+    - name: Static Analysis
+      run: ./vendor/bin/psalm --update-baseline --no-cache
```
Comment (mlahargou) [inline .github/workflows/php.yml:61]: Why did we add `update-baseline` here?

File: .github/workflows/php.yml [hunk 5]
```diff
@@ -0,0 +1,64 @@
+name: PHP Composer
+
+on:
+  push:
+    branches: [ master ]
```
Comment (mlahargou) [inline .github/workflows/php.yml:5]: I thought this would only trigger the workflow on master. 

https://docs.github.com/en/actions/learn-github-actions/workflow-syntax-for-github-actions#example-including-branches-and-tags

**Comment by mlahargou** [comment]
CR 📱  with a couple of questions/comments.

File: .github/workflows/php.yml (hunk 1 above)
Comment (davidrans) [reply .github/workflows/php.yml:40]: It replaces `services: memcached`. 

Installing PHP with the `memcached` extension takes care of the `pecl install` bit I think.

File: .github/workflows/php.yml (hunk 2 above)
Comment (davidrans) [reply .github/workflows/php.yml:45]: I think the OS is in the cache key, so that it _doesn't_ get cached across different OS's.

I'm not sure if it caches just across different steps, or also subsequent CI runs? I stole this from the PHP CI template, and forgot to look into it more.

File: .github/workflows/php.yml (hunk 4 above)
Comment (davidrans) [reply .github/workflows/php.yml:61]: I think we want this to fail if the baseline changes. But I didn't actually check that that's what happens here. Might need to manually return a non-zero exit code if there's a diff.

File: .github/workflows/php.yml (hunk 3 above)
Comment (davidrans) [reply .github/workflows/php.yml:21]: Good question. I modeled this after the PHPUnit repo, and this is what they were using.

I'm not sure what would update in a newer version of a Linux distro that would cause this to fail, so I'm not sure if we would want to fail tests on it or not.

File: .github/workflows/php.yml [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+  pull_request:
+    branches: [ master ]
```
Comment (bcooper94) [inline .github/workflows/php.yml:7]: Similar to @mlahargou's comment above, does this mean we're only running this workflow on pull requests that we want to merge into master? If we have two different PRs with the second branching off of the first, and set the second PR to merge into the first PR's branch, will this run for the second PR?

**Comment by bcooper94** [comment]
CR 👍 but deploy_block ⚡ on a few questions.

File: .github/workflows/php.yml (hunk 6 above)
Comment (davidrans) [reply .github/workflows/php.yml]: Good call 0c9e702

File: .github/workflows/php.yml (hunk 5 above)
Comment (davidrans) [reply .github/workflows/php.yml:5]: After 0c9e702, I think this ought to trigger when pushing to `master` or when pushing to any pull request.

File: .github/workflows/php.yml (hunk 4 above)
Comment (davidrans) [reply .github/workflows/php.yml]: I changed this back 86c1a7b

**Comment by davidrans** [comment]
un_deploy_block 👍

File: .github/workflows/php.yml (hunk 2 above)
Comment (sctice-ifixit) [reply .github/workflows/php.yml:44]: https://github.com/actions/cache

I think we're caching the vendor directory (i.e., everything installed by composer), and only invalidating the cache if a composer.lock file (in any directory) changes (different hash), or if the OS changes. In the event of a hit, we restore the vendor directory from the cache. Caching is scoped to the current branch, but if that's a miss, it will try the parent branch, so if a branch doesn't change composer.lock, then I expect it to get the vendor dir from its parent (probably master). The `restore-keys` argument is used if there's a miss on `key`, and it's a most-to-least-specific list of key prefixes to try. So even if there isn't a hit for the current composer.lock, this config will pull the most recent cached value from the branch scope, falling back to the parent branch scope.

Basically, we're going to try really hard to get the vendor dir from _some_ previous build, and then we're going to run `composer install` on top of that. We have the option to skip a job based on a previous cache job resulting in a hit, but we wouldn't want that here because our key config means that we can get a vendor dir that doesn't fully satisfy our composer.lock.

There's a 10G per repo limit on total cache size, and items are evicted after 7 days without access.

File: .github/workflows/php.yml [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+      with:
+        path: vendor
+        key: ${{ runner.os }}-php-${{ hashFiles('**/composer.lock') }}
+        restore-keys: |
+          ${{ runner.os }}-php-
+
+    - name: Install dependencies
+      run: composer install --prefer-dist --no-progress
+
+    - name: Run tests with phpunit
```
Comment (sctice-ifixit) [inline .github/workflows/php.yml:56]: Was there a reason to switch to running PHPUnit before Psalm? It makes sense to me to do the static checks first.

**Comment by sctice-ifixit** [comment]
CR :flags: Nice!

I don't see a reason to change the order of Psalm and unit tests, but I also don't think it's important enough to block on.

File: .github/workflows/php.yml (hunk 2 above)
Comment (danielbeardsley) [reply .github/workflows/php.yml:44]: Thanks for always being thorough and informative Shawn!

**Comment by danielbeardsley** [comment]
CR :+1: 

This looks great!  Is there any way to test it?

File: .github/workflows/php.yml (hunk 7 above)
Comment (davidrans) [reply .github/workflows/php.yml:56]: There was no reason. They both take trivial amounts of time, so I don't think it matters:
image

**Comment by davidrans** [comment]
> Is there any way to test it?

Test that the actions are running?
image

**Comment by danielbeardsley** [comment]
Wuuh?! I thought we didn't have Github Actions available on any repos of ours?

**Comment by davidrans** [comment]
> Wuuh?! I thought we didn't have Github Actions available on any repos of ours?

It works on public repos 👍
